File: valo_tui/screens/events.py
"""[e] events — the parent hub. Browse VCT / Challenger stages and pick one to
focus the whole app on it. Enter on a row opens that event's overview."""

from __future__ import annotations

from rich.text import Text

from textual.app import ComposeResult
from textual.containers import Vertical
from textual.widgets import Label

from ..data import cache
from ..data.models import EventCard
from .widgets import ACCENT, LIVE, MUTED, TEXT, VimDataTable

# Ongoing first, then upcoming, then completed.
_ORDER = {"ongoing": 0, "upcoming": 1, "completed": 2}


def _status_cell(status: str) -> Text:
    s = (status or "").lower()
    if s.startswith("ongo"):
        return Text("● live", style=LIVE)
    if s.startswith("upcom"):
        return Text("○ soon", style=MUTED)
    return Text("✓ done", style=MUTED)


class EventsView(Vertical):
    def compose(self) -> ComposeResult:
        yield Label("events", classes="page-title")
        yield Label("enter → open event · j/k move", classes="hint")
        yield VimDataTable(id="events-table", cursor_type="row", zebra_stripes=False)

    def on_mount(self) -> None:
        t = self.query_one(VimDataTable)
        t.add_column("status", width=8, key="status")
        t.add_column("event", width=40, key="event")
        t.add_column("region", width=12, key="region")
        t.add_column("when", key="when")
        self.load_data()

    def load_data(self) -> None:
        t = self.query_one(VimDataTable)
        prev = t.cursor_row
        t.clear()
        events = sorted(
            cache.active_events(),
            key=lambda e: _ORDER.get((e.status or "").lower(), 9),
        )
        for e in events:
            t.add_row(*self._row(e), key=str(e.id))
        if events:
            t.move_cursor(row=min(prev, len(events) - 1))
        else:
            t.add_row(Text(""), Text("no events in cache — start the worker", style=MUTED),
                      Text(""), Text(""))

    def _row(self, e: EventCard) -> tuple[Text, ...]:
        when = ""
        if e.start and e.end:
            when = f"{e.start} – {e.end}"
        elif e.start:
            when = e.start
        return (
            _status_cell(e.status),
            Text(e.name, style=TEXT),
            Text(e.region or "—", style=ACCENT if e.region else MUTED),
            Text(when, style=MUTED),
        )

File: internal/screens/events.go
package screens

import (
	"fmt"
	"sort"
	"strings"

	tea "charm.land/bubbletea/v2"

	"github.com/jashkarangiya/valo-tui/internal/data"
	"github.com/jashkarangiya/valo-tui/internal/widgets"
)

var eventOrder = map[string]int{"ongoing": 0, "upcoming": 1, "completed": 2}

// Events is the [e] hub: browse tournaments and pick one to focus the app on.
type Events struct {
	w, h  int
	table widgets.Table
}

func NewEvents(w, h int) Events {
	e := Events{w: w, h: h, table: widgets.NewTable(
		widgets.Column{Title: "status", Width: 8},
		widgets.Column{Title: "event", Width: 40},
		widgets.Column{Title: "region", Width: 12},
		widgets.Column{Title: "when", Width: 20},
	)}
	e.Load()
	return e
}

func (s *Events) SetSize(w, h int) {
	s.w, s.h = w, h
	s.table.SetHeight(h - 4)
}

func (s *Events) Load() {
	events := data.ActiveEvents()
	sort.SliceStable(events, func(a, b int) bool {
		return eventOrder[strings.ToLower(events[a].Status)] < eventOrder[strings.ToLower(events[b].Status)]
	})
	rows := make([]widgets.Row, 0, len(events))
	for _, e := range events {
		when := ""
		switch {
		case e.Start != "" && e.End != "":
			when = e.Start + " – " + e.End
		case e.Start != "":
			when = e.Start
		}
		region := e.Region
		regionStyle := accentSt
		if region == "" {
			region = "—"
			regionStyle = mutedSt
		}
		rows = append(rows, widgets.Row{
			Key: fmt.Sprint(e.ID),
			Cells: []widgets.Cell{
				statusCell(e.Status),
				{Text: e.Name, Style: textSt},
				{Text: region, Style: regionStyle},
				{Text: when, Style: mutedSt},
			},
		})
	}
	s.table.SetRows(rows)
}

func statusCell(status string) widgets.Cell {
	s := strings.ToLower(status)
	switch {
	case strings.HasPrefix(s, "ongo"):
		return widgets.Cell{Text: "● live", Style: liveSt}
	case strings.HasPrefix(s, "upcom"):
		return widgets.Cell{Text: "○ soon", Style: mutedSt}
	default:
		return widgets.Cell{Text: "✓ done", Style: mutedSt}
	}
}

func (s *Events) Focus()                           { s.table.Focus() }
func (s *Events) Blur()                            { s.table.Blur() }
func (s Events) Selected() string                  { return s.table.SelectedKey() }
func (s *Events) ClickVisual(i int) (string, bool) { return s.table.ClickVisual(i) }

func (s Events) Update(msg tea.Msg) (Events, tea.Cmd) {
	if k, ok := msg.(tea.KeyPressMsg); ok {
		switch k.String() {
		case "j", "down":
			s.table.MoveCursor(1)
		case "k", "up":
			s.table.MoveCursor(-1)
		}
	}
	return s, nil
}

func (s Events) View() string {
	header := title("events") + "\n" + hint("enter → open event · j/k move") + "\n\n"
	if s.table.Len() == 0 {
		return header + muted("no events in cache — seed it (see README)")
	}
	return header + s.table.View()
}
